File: ch3/UVa00750.ml
open Printf
open Scanf

let row = Array.make 8 0
let a = ref 0
let b = ref 0
let line_counter = ref 0

let can_place r c =
  List.init c (fun i -> i) |> List.for_all (fun prev ->
      row.(prev) <> r && abs (row.(prev) - r) <> abs (prev - c))

let rec backtrack c =
  if c = 8 then begin
    line_counter := !line_counter+1;
    printf "%2d      %d" !line_counter (row.(0)+1);
    for j = 1 to 7 do
      printf " %d" (row.(j)+1);
    done;
    printf "\n"
  end
  else
    for r = 0 to 7 do
      if (c = !b) = (r = !a) && can_place r c then begin
        row.(c) <- r; backtrack (c+1)
      end
    done

let () =
  let tc = scanf "%d\n" (fun x -> x) in
  for i = 1 to tc do
    scanf "\n%d %d\n" (fun x y -> a := x-1; b := y-1);
    line_counter := 0;
    printf "SOLN       COLUMN\n";
    printf " #      1 2 3 4 5 6 7 8\n\n";
    backtrack 0;
    if i < tc then printf "\n"
  done

File: ch3/cs/UVa00750.java
import java.util.*;

class Main { /* UVa 00750 - 8 Queens Chess Problem */
  private static int[] row = new int[9];
  private static int TC, a, b, lineCounter; // it is ok to use global variables in competitive programming

  private static boolean place(int col, int tryrow) {
    for (int prev = 1; prev < col; prev++) // check previously placed queens
      if (row[prev] == tryrow || (Math.abs(row[prev] - tryrow) == Math.abs(prev - col)))
        return false; // an infeasible solution if share same row or same diagonal
    return true;
  }

  private static void backtrack(int col) {
    for (int tryrow = 1; tryrow <= 8; tryrow++) // try all possible row
      if (place(col, tryrow)) { // if can place a queen at this col and row...
        row[col] = tryrow; // put this queen in this col and row
        if (col == 8 && row[b] == a) { // a candidate solution & (a, b) has 1 queen
          System.out.printf("%2d      %d", ++lineCounter, row[1]);
          for (int j = 2; j <= 8; j++) System.out.printf(" %d", row[j]);
          System.out.printf("\n");
        }
        else
          backtrack(col + 1); // recursively try next column
  }   }
  
  public static void main(String[] args) {
    Scanner sc = new Scanner(System.in);
    TC = sc.nextInt();
    while (TC-- > 0) {
      a = sc.nextInt();
      b = sc.nextInt();
      for (int i = 0; i < 9; i++) row[i] = 0;
      lineCounter = 0;
      System.out.printf("SOLN       COLUMN\n");
      System.out.printf(" #      1 2 3 4 5 6 7 8\n\n");
      backtrack(1); // generate all possible 8! candidate solutions
      if (TC > 0) System.out.printf("\n");
    }
  }
}
